common/dynamicconfig/memory_client.go

```diff
@@ -29,32 +29,40 @@ import (
 )
 
 type MemoryClient struct {
-	sync.RWMutex
+	lock      sync.RWMutex
 	overrides map[Key]any
 	fallback  Client
 }
 
 func (d *MemoryClient) GetRawValue(name Key) (any, bool) {
-	d.RLock()
-	defer d.RUnlock()
+	d.lock.RLock()
+	defer d.lock.RUnlock()
 	v, ok := d.overrides[name]
+	if value, ok := v.([]ConstrainedValue); ok {
+		return value[0].Value, true
+	}
 	return v, ok
 }
 
 func (d *MemoryClient) GetValue(name Key) []ConstrainedValue {
-	if val, ok := d.GetRawValue(name); ok {
-		return []ConstrainedValue{{Value: val}}
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	if v, ok := d.overrides[name]; ok {
+		if value, ok := v.([]ConstrainedValue); ok {
+			return value
+		}
+		return []ConstrainedValue{{Value: v}}
 	}
 	return d.fallback.GetValue(name)
```

Comment:
Should probably not hold the lock while calling the fallback? I guess it's unlikely to matter in practice.

But I'm also still questioning the need for the fallback. Specifically: the next step is to make MemoryClient implement NotifyingClient so subscriptions work instantaneously. Doing that with fallback is a lot more complicated since you have to check if that one can notify, subscribe to it, pass through callbacks, etc. And having a fallback but having subscriptions only work on MemoryClient but not the fallback is confusing semantics. So I might still lean towards dropping it, if there's not a very clear use case.

Reply:
Right. Now I think it is better to remove fallback.